File: examples/integrator_example.py
#!/usr/bin/env python3

# Copyright 2024 Universidad Politécnica de Madrid
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#    * Redistributions of source code must retain the above copyright
#      notice, this list of conditions and the following disclaimer.
#
#    * Redistributions in binary form must reproduce the above copyright
#      notice, this list of conditions and the following disclaimer in the
#      documentation and/or other materials provided with the distribution.
#
#    * Neither the name of the Universidad Politécnica de Madrid nor the names of its
#      contributors may be used to endorse or promote products derived from
#      this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

"""MPC Test using Acados Integrator."""

__authors__ = 'Rafael Pérez Seguí'
__copyright__ = 'Copyright (c) 2022 Universidad Politécnica de Madrid'
__license__ = 'BSD-3-Clause'

from functools import wraps

import utils.sym_example_utils as utils
from mpc.mpc_controller import MPC, mpc_lib
import numpy as np
from tqdm import tqdm


def trajectory_point_to_mpc_reference(trajectory_point):
    """Convert trajectory point to MPC reference."""
    ref_position, ref_velocity, _, ref_yaw = \
        trajectory_point
    return mpc_lib.CaState.get_state(
        position=ref_position,
        orientation=utils.euler_to_quaternion(0.0, 0.0, ref_yaw),
        linear_velocity=ref_velocity
    )


def progress_bar(func):
    @wraps(func)
    def wrapper(logger, mpc, simulator, trajectory_generator, yaml_data, *args, **kwargs):
        sim_max_t = yaml_data.sim_time

        pbar = tqdm(total=sim_max_t, desc=f'Progress {func.__name__}', unit='iter',
                    bar_format='{l_bar}{bar} | {n:.4f}/{total:.2f} '
                    '[{elapsed}<{remaining}, {rate_fmt}]')

        result = func(logger, mpc, simulator, trajectory_generator, yaml_data, pbar, *args, **kwargs)

        pbar.close()
        return result
    return wrapper


@progress_bar
def test_trajectory_controller(
        logger: utils.CsvLogger,
        mpc: MPC,
        simulator,
        trajectory_generator,
        yaml_data: utils.YamlData,
        pbar):
    """Test trajectory controller."""

    x = mpc_lib.CaState.get_state()
    y = mpc_lib.CaState.get_state()
    u = mpc_lib.CaControl.get_control()

    prediction_steps = mpc.prediction_steps
    prediction_horizon = mpc.prediction_horizon
    tf = prediction_horizon / prediction_steps

    t = 0.0
    min_time = trajectory_generator.get_min_time()
    max_time = trajectory_generator.get_max_time()

    logger.save(t, x, y , u)
    while t < yaml_data.sim_time:
        t_eval = t
        reference_trajectory = np.zeros((prediction_steps+1, mpc.x_dim))
        # print(reference_trajectory.shape)  # (101, 10)
        for i in range(prediction_steps+1):
            if t_eval >= max_time:
                t_eval = max_time
            elif t_eval <= min_time:
                t_eval = min_time
            trajectory_point = trajectory_generator.evaluate_trajectory(t_eval)
            reference_trajectory[i, :] = trajectory_point_to_mpc_reference(trajectory_point)
            t_eval += tf
        # u = mpc.compute_control_action(x, reference_trajectory[:-1], reference_trajectory[-1][:mpc.x_dim])
        u = mpc.evaluate(x, reference_trajectory[:-1], reference_trajectory[-1][:mpc.x_dim])

        simulator.set('x', x)
        simulator.set('u', u)
        status = simulator.solve()
        if status != 0:
            raise Exception(
                'acados integrator returned status {}. Exiting.'.format(status))
        x = simulator.get("x")

        t += yaml_data.dt

        y = mpc_lib.CaState.get_state(
            position=reference_trajectory[0][0:3],
            orientation=reference_trajectory[0][3:7],
            linear_velocity=reference_trajectory[0][7:10])
        logger.save(t, x, y, u)

        pbar.update(yaml_data.dt)

    logger.close()


if __name__ == '__main__':
    # Params
    yaml_data = utils.read_yaml_params('examples/integrator_simulation_config.yaml')

    # Logger
    file_name = 'mpc_log.csv'
    logger = utils.CsvLogger(file_name)

    # MPC
    mpc_params = mpc_lib.AcadosMPCParams(
        Q=yaml_data.mpc_gains.Q,
        Qe=yaml_data.mpc_gains.Qe,
        R=yaml_data.mpc_gains.R,
        lbu=yaml_data.mpc_gains.lbu,
        ubu=yaml_data.mpc_gains.ubu,
        p=yaml_data.mpc_gains.p
    )
    mpc = MPC(
        prediction_steps=100,
        prediction_horizon=0.5,
        params=mpc_params
    )

    # Simulator
    Tf = mpc.prediction_horizon
    N = mpc.prediction_steps
    dt = Tf / N
    simulator = mpc.export_integrador(dt)

    # Trajectory generator
    trajectory_generator = utils.get_trajectory_generator(
        initial_position=np.zeros(3),
        waypoints=yaml_data.waypoints,
        speed=yaml_data.trajectory_generator_max_speed
    )
    trajectory_generator.set_path_facing(yaml_data.path_facing)
    
    test_trajectory_controller(
        logger,
        mpc,
        simulator,
        trajectory_generator,
        yaml_data)

File: examples/integrator_example.cpp
// Copyright 2024 Universidad Politécnica de Madrid
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//    * Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//    * Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in the
//      documentation and/or other materials provided with the distribution.
//
//    * Neither the name of the Universidad Politécnica de Madrid nor the names
//    of its
//      contributors may be used to endorse or promote products derived from
//      this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

/**
 * @file integrator_example.cpp
 *
 * Acados MPC examples using Acados Sim solver.
 *
 * @author Rafael Perez-Segui <r.psegui@upm.es>
 */

#include <yaml-cpp/yaml.h>
#include <chrono>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <memory>
#include <stdexcept>

#include <dynamic_trajectory_generator/dynamic_trajectory.hpp>
#include <dynamic_trajectory_generator/dynamic_waypoint.hpp>

#include "acados_mpc/acados_mpc.hpp"
#include "acados_mpc/acados_sim_solver.hpp"

#include "utils/sym_example_utils.hpp"

namespace acados_mpc {
namespace acados_mpc_examples {

using DynamicTrajectory = dynamic_traj_generator::DynamicTrajectory;
using DynamicWaypoint   = dynamic_traj_generator::DynamicWaypoint;

void traj_generator_ref_to_mpc_ref(const dynamic_traj_generator::References &references,
                                   MPCData *mpc_data,
                                   int index) {
  if (index == MPC_N) {
    mpc_data->reference_end.set_data(0, references.position.x());
    mpc_data->reference_end.set_data(1, references.position.y());
    mpc_data->reference_end.set_data(2, references.position.z());
    mpc_data->reference_end.set_data(3, 1.0);
    mpc_data->reference_end.set_data(4, 0.0);
    mpc_data->reference_end.set_data(5, 0.0);
    mpc_data->reference_end.set_data(6, 0.0);
    mpc_data->reference_end.set_data(7, references.velocity.x());
    mpc_data->reference_end.set_data(8, references.velocity.y());
    mpc_data->reference_end.set_data(9, references.velocity.z());
    return;
  } else if (index > MPC_N) {
    throw std::out_of_range("Index out of range.");
  }
  mpc_data->reference.set_data(index, 0, references.position.x());
  mpc_data->reference.set_data(index, 1, references.position.y());
  mpc_data->reference.set_data(index, 2, references.position.z());
  mpc_data->reference.set_data(index, 3, 1.0);
  mpc_data->reference.set_data(index, 4, 0.0);
  mpc_data->reference.set_data(index, 5, 0.0);
  mpc_data->reference.set_data(index, 6, 0.0);
  mpc_data->reference.set_data(index, 7, references.velocity.x());
  mpc_data->reference.set_data(index, 8, references.velocity.y());
  mpc_data->reference.set_data(index, 9, references.velocity.z());
}

void test_mpc_controller(CsvLogger &logger,
                         MPC &mpc,
                         MPCSimSolver &simulator,
                         std::unique_ptr<DynamicTrajectory> &trajectory_generator,
                         YamlData &yaml_data) {
  // Initialize dynamic trajectory generator
  dynamic_traj_generator::References references;
  double tg_max_time = trajectory_generator->getMaxTime();

  // MPC Parameters
  MPCData *mpc_data       = mpc.get_data();
  double prediction_steps = mpc.get_prediction_steps();
  // double prediction_horizon = mpc.get_prediction_horizon();
  double tf = mpc.get_prediction_time_step();

  // Simulation
  double min_time = trajectory_generator->getMinTime();  // seconds
  double t        = 0.0;                                 // seconds
  logger.save(t, mpc_data);

  // Time measurement
  const std::size_t n_iterations = static_cast<std::size_t>(yaml_data.sim_time / yaml_data.dt);
  std::vector<double> mpc_times;
  mpc_times.reserve(n_iterations);
  std::vector<double> sim_times;
  sim_times.reserve(n_iterations);
  std::vector<double> total_times;
  total_times.reserve(n_iterations);

  for (double t = 0; t < yaml_data.sim_time; t += yaml_data.dt) {
    auto iter_start = std::chrono::high_resolution_clock::now();

    double t_eval = t;

    // yref from 0 to N-1 (N steps) and yref_N from N to N
    for (int i = 0; i < prediction_steps + 1; i++) {
      if (t_eval >= tg_max_time) {
        t_eval = tg_max_time - yaml_data.dt;
      } else if (t_eval <= min_time) {
        t_eval = min_time;
      }
      trajectory_generator->evaluateTrajectory(t_eval, references);
      traj_generator_ref_to_mpc_ref(references, mpc_data, i);
      t_eval += tf;
    }

    // Solve MPC
    auto mpc_start = std::chrono::high_resolution_clock::now();
    mpc.solve();
    auto mpc_end = std::chrono::high_resolution_clock::now();

    // Simulate
    auto sim_start = std::chrono::high_resolution_clock::now();
    simulator.solve(mpc_data);
    auto sim_end = std::chrono::high_resolution_clock::now();

    std::chrono::duration<double> mpc_duration   = mpc_end - mpc_start;
    std::chrono::duration<double> sim_duration   = sim_end - sim_start;
    std::chrono::duration<double> total_duration = sim_end - iter_start;
    mpc_times.push_back(mpc_duration.count());
    sim_times.push_back(sim_duration.count());
    total_times.push_back(total_duration.count());

    logger.save(t, mpc_data);
  }
  logger.close();

  // Print time measurements and its average
  double mpc_avg_time = std::accumulate(mpc_times.begin(), mpc_times.end(), 0.0) / mpc_times.size();
  double sim_avg_time = std::accumulate(sim_times.begin(), sim_times.end(), 0.0) / sim_times.size();
  double total_avg_time =
      std::accumulate(total_times.begin(), total_times.end(), 0.0) / total_times.size();
  std::cout << "MPC average time: " << mpc_avg_time << " s" << std::endl;
  std::cout << "Simulator average time: " << sim_avg_time << " s" << std::endl;
  std::cout << "Total average time: " << total_avg_time << " s" << std::endl;
}

}  // namespace acados_mpc_examples
}  // namespace acados_mpc

int main(int argc, char **argv) {
  // Logger
  std::string file_name = "mpc_log.csv";
  acados_mpc::acados_mpc_examples::CsvLogger logger(file_name);

  // Params
  acados_mpc::acados_mpc_examples::YamlData yaml_data;
  acados_mpc::acados_mpc_examples::read_yaml_params("examples/integrator_simulation_config.yaml",
                                                    yaml_data);

  // Initialize MPC
  acados_mpc::MPC mpc = acados_mpc::MPC();

  // Update MPC gains and bounds
  mpc.get_gains()->set_Q(yaml_data.mpc_data.Q);
  mpc.get_gains()->set_Q_end(yaml_data.mpc_data.Qe);
  mpc.get_gains()->set_R(yaml_data.mpc_data.R);
  mpc.get_bounds()->set_lbu(yaml_data.mpc_data.lbu);
  mpc.get_bounds()->set_ubu(yaml_data.mpc_data.ubu);
  mpc.get_online_params()->set_data(yaml_data.mpc_data.p);
  mpc.update_bounds();
  mpc.update_gains();
  mpc.update_online_params();

  // Initialize integrator
  acados_mpc::MPCSimSolver simulator = acados_mpc::MPCSimSolver();

  // Initialize trajectory generator
  auto trajectory_generator = acados_mpc::acados_mpc_examples::get_trajectory_generator(
      Eigen::Vector3d::Zero(), yaml_data.waypoints, yaml_data.trajectory_generator_max_speed);

  acados_mpc::acados_mpc_examples::test_mpc_controller(logger, mpc, simulator, trajectory_generator,
                                                       yaml_data);
  return 0;
}
